src/v/cluster/state_machine_registry.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+#include "raft/consensus.h"
+#include "raft/fwd.h"
+#include "storage/ntp_config.h"
+
+namespace cluster {
+
+/**
+ * State machine factory is a class used by registry to create stm instance if
+ * it is required for a given Raft group. The factory has two main
+ * responsibilities expressed by its interface method.
+ * Firstly it has to answer query if given stm is required for particular raft
+ * group. Secondly it must created an stm instance using a builder which is
+ * passed into `create` method. State machine factory must encapsulate all
+ * dependencies required to create state machine
+ */
+struct state_machine_factory {
+    /**
+     * Must return true if STM should be created for a partition underlaid by
+     * passed raft group
+     */
+    virtual bool is_applicable_for(const storage::ntp_config&) const = 0;
+
+    /**
+     * A method must call builder interface to create STM instance.
+     */
+    virtual void create(raft::state_machine_manager_builder&, raft::consensus*)
+      = 0;
+
+    virtual ~state_machine_factory() = default;
+};
+
+/**
+ * State machine registry is a class holding all registered state machines
+ * factories. Registry is used by partition_manger whenever a new partition
+ * instance is created. Registry builds all state machines that are required for
+ * a give partition instance.
+ */
+class state_machine_registry {
+public:
+    template<typename T, typename... Args>
+    void register_factory(Args&&... args) {
+        _stm_factories.push_back(
+          std::make_unique<T>(std::forward<Args>(args)...));
+    }
+
+    raft::state_machine_manager_builder
+    make_builder_for(raft::consensus* raft) {
+        raft::state_machine_manager_builder builder;
+        for (auto& factory : _stm_factories) {
+            if (factory->is_applicable_for(raft->log_config())) {
+                factory->create(builder, raft);
+            }
+        }
```

Comment:
Is there a benefit to having the state machine registry manage and invoke factories? It seems like another model would be to deliver partition life cycle events to each registered module which could itself build the STM however it wanted without factory abstractions.

I guess ultimately it's not that much different. But what if a module wanted to track all of the partitions that its STMs have been attached to? Would that module need to capture references to the module in the factory closure that would reach back into the module to record the event? What would be the proper order for shutdown?

Reply:
I was considering the event based mechanism before but i found it complex. With the events based mechanism we would introduce a dependency on the startup order where notification upcall would have to be registered before the `partition_manager` starts.

Reply:
>notification upcall would have to be registered before the partition_manager starts.

would this be through the existing notification system we have, or be a notification through the machinery of STMs? If the former, then it seems complicated because it's an entirely separate use case for notifications and we'd need even duplicate all the logic to decide if the partition would be registered through the separate STM machinery. In this case it would be far cleaner / clearer for the module to capture a reference to itself in the factory.

I don't think we need to go changing anything the strategy at this point, but using the notification subsystem seems like suboptimal for solving this problem if it were to pop up in the future.